### ISSUE TITLE
Are all the training data used in each epoch please?

### ISSUE THREAD

**Issue by AliShahin** [ftramer/Handcrafted-DP#2]
Dear Florian,

It seems your code does not use all the training data in each epoch, see https://github.com/ftramer/Handcrafted-DP/blob/main/train_utils.py#L21, or I am missing something?

If yes, does it affect the privacy of samples please?

Thanks,
Ali

**Comment by ftramer** [comment]
That's right!

The analysis of DP-SGD is based on the number of total training steps, where each training step samples a batch by picking each example with probability B/N (where N is the number of training examples, and B is the expected batch size).
This is usually called "Poisson sampling". There's thus no concept of "epoch" in the formal analysis of DP-SGD.

Now, in practice we "cheat" by going through the entire training set with batches of size exactly B. This isn't strictly equivalent to Poisson sampling but it is easier to integrate with our standard way of training ML models (and, empirically, it performs essentially the same way as Poisson sampling).
What the code in this repository does, is that it splits the training set into batches of size exactly B, and if there's a last batch of size <B, that last batch is dropped.

**Comment by AliShahin** [comment]
Thanks for your prompt reply!

The above B that you mentioned is the mini_batch_size in your code, not the batch_size, right? I assume batch_size in your code is for the parallel computation to speed up the training (?)

If B is the mini_batch_size, you might remove more than 1 mini_batch of size B due to the n_acc_steps https://github.com/ftramer/Handcrafted-DP/blob/main/train_utils.py#L19. I am not sure that I understand what is n_acc_steps for. For example, in CIFAR-10, the mini_batch_size is 256 and the batch_size is 1024, so n_acc_steps is 4 based on https://github.com/ftramer/Handcrafted-DP/blob/main/cnns.py#L36

**Comment by ftramer** [comment]
Yes, things get a bit confusing here in terms of the terminology, sorry.

- the `batch_size` parameter that you pass to cnns.py is the parameter `B` above. It determines how many examples get aggregated into a gradient before that gradient is noised. This is the parameter that matters for the analysis of DP-SGD.

- the batch size `B` might be too large to fit into memory. For this reason, we introduce a `mini_batch_size` (default of 256). During training, we continuously load batches of size `mini_batch_size`, compute per-example gradients, and store the summed gradient. After doing this for `n_acc_steps = batch_size / mini_batch_size` steps, we've accumulated gradients for `B` examples and we actually compute the DP-SGD update step.

- Now back to the removal of the last batch in train_utils.py: suppose we have `50000` points, and that `virtual_batch_size=256` and `batch_size=1024`. So `n_acc_steps=4`. Our data loader will yield 195 batches of size 256, and 1 batch of size 80. The last 4 mini batches can thus be ignored, as those will yield a full batch of size `3*256+80 = 848 < 1024`.

**Comment by AliShahin** [comment]
Thanks for the clarification! 

What do you think of considering also the last batch of 848 data points by adding noise scaled to 848 as opposed to 1024 please?

**Comment by ftramer** [comment]
You can definitely do that. 
I'm not sure if Opacus has the tooling implemented to track updates of DP-SGD with different sampling rates but it shouldn't be too hard to add that functionality.
